src/ws/resharder.ts

```diff
@@ -1,31 +1,141 @@
 import { GetGatewayBot } from "../types/gateway/getGatewayBot.ts";
-import { GatewayManager } from "./gateway_manager.ts";
+import { DiscordReady } from "../types/gateway/ready.ts";
+import { Collection } from "../util/collection.ts";
+import { createGatewayManager, GatewayManager } from "./gateway_manager.ts";
 
 /** The handler to automatically reshard when necessary. */
-export async function resharder(gateway: GatewayManager) {
-  // TODO: is it possible to route this to REST?
-  const results = (await fetch(`https://discord.com/api/gateway/bot`, {
-    headers: { Authorization: `Bot ${gateway.token}` },
-  }).then((res) => res.json())) as GetGatewayBot;
+export async function resharder(
+  oldGateway: GatewayManager,
+  results: GetGatewayBot,
+) {
+  oldGateway.debug("[Resharding] Starting the reshard process.");
 
-  const percentage = ((results.shards - gateway.maxShards) / gateway.maxShards) * 100;
-  // Less than necessary% being used so do nothing
-  if (percentage < gateway.reshardPercentage) return;
-
-  // Don't have enough identify rate limits to reshard
-  if (results.sessionStartLimit.remaining < results.shards) {
-    return;
-  }
+  const gateway = createGatewayManager({
+    ...oldGateway,
+    // IGNORE EVENTS FOR NOW
+    handleDiscordPayload: async function (_, data, shardId) {
+      if (data.t === "READY") {
+        const payload = data.d as DiscordReady;
+        for (const guild of payload.guilds) {
+          gateway.resharding.markNewGuildShardId(BigInt(guild.id), shardId);
+        }
```

Comment:
instead of having this a function which updates only one guild id, it should take an array of guild ids and the new shard id